File: typescript/packages/core/src/vfs/gslides/gslides.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { BoundVFS } from '../bound.ts'
import { GSLIDES_IO } from '../../commands/builtin/gslides/io.ts'
import { GSlidesAccessor } from '../../accessor/gslides.ts'

import { GSLIDES_COMMANDS } from '../../commands/builtin/gslides/index.ts'
import type { RegisteredCommand } from '../../commands/config.ts'
import { TokenManager } from '../../core/google/client.ts'

import { GSLIDES_OPS } from '../../ops/gslides/index.ts'
import type { RegisteredOp } from '../../ops/registry.ts'

import type { VFS } from '../base.ts'
import { GSLIDES_PROMPT, GSLIDES_WRITE_PROMPT } from './prompt.ts'
import { VFSName } from '../../types.ts'

import { redactGSlidesConfig, type GSlidesConfig, type GSlidesConfigRedacted } from './config.ts'

export interface GSlidesVFSState {
  type: string
  config: GSlidesConfigRedacted
}

export class GSlidesVFS extends BoundVFS<GSlidesAccessor> implements VFS {
  readonly kind: string = VFSName.GSLIDES
  readonly cachesReads: boolean = true
  override readonly indexTtl: number = 86_400
  readonly prompt: string = GSLIDES_PROMPT
  readonly writePrompt: string = GSLIDES_WRITE_PROMPT
  readonly config: GSlidesConfig
  readonly accessor: GSlidesAccessor

  constructor(config: GSlidesConfig) {
    super(GSLIDES_IO)
    this.config = config
    const tm = new TokenManager(config)
    this.accessor = new GSlidesAccessor({ tokenManager: tm })
  }

  commands(): readonly RegisteredCommand[] {
    return GSLIDES_COMMANDS
  }

  ops(): readonly RegisteredOp[] {
    return GSLIDES_OPS
  }

  override getState(): Promise<GSlidesVFSState> {
    return Promise.resolve({
      type: this.kind,
      config: redactGSlidesConfig(this.config),
    })
  }
}

File: python/mirage/vfs/gslides/gslides.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from typing import Any

from mirage.accessor.gslides import GSlidesAccessor
from mirage.commands.builtin.gslides import COMMANDS
from mirage.commands.builtin.gslides.io import IO
from mirage.core.google.client import TokenManager
from mirage.ops.gslides import OPS as GSLIDES_VFS_OPS
from mirage.types import VFSName
from mirage.vfs.bound import BoundVFS
from mirage.vfs.gslides.config import GSlidesConfig
from mirage.vfs.gslides.prompt import PROMPT, WRITE_PROMPT


class GSlidesVFS(BoundVFS):

    accessor: GSlidesAccessor
    name: str = VFSName.GSLIDES
    caches_reads: bool = True
    # An API-backed tree that changes rarely; a day-long index spares the
    # provider a full re-walk every 10 minutes. Mirrors the TypeScript
    # VFS.
    index_ttl: float = 86_400
    PROMPT: str = PROMPT
    WRITE_PROMPT: str = WRITE_PROMPT

    def __init__(self, config: GSlidesConfig) -> None:
        super().__init__(io=IO)
        self.config = config
        self._token_manager = TokenManager(config)
        self.accessor = GSlidesAccessor(self.config, self._token_manager)

        for fn in COMMANDS:
            self.register(fn)
        for fn in GSLIDES_VFS_OPS:
            self.register_op(fn)

    async def close(self) -> None:
        """Drain the token manager's connection pool with the VFS."""
        await self._token_manager.close()
        await super().close()

    def get_state(self) -> dict[str, Any]:
        return self.config_state(self.config)
